### DISCUSSION TITLE
iOS Shortcuts

### DISCUSSION THREAD

**Discussion by kejadlen** [babybuddy/babybuddy#300] [Show and tell]
A few basic iOS shortcuts that can be used as templates if other people want to use Siri to add diaper changes and start/stop timers. I’ve only implemented the functionality I need, but putting them here for now so other people can pull them down and configure it for their own needs.

- [Add Diaper Change](https://www.icloud.com/shortcuts/10824127057043eeb7e1edc8fb7ad3e9)
- [Add Diaper Change](https://www.icloud.com/shortcuts/3e7baf6586ac43af987614f51a7bcb66) (with color support for solids by @azosnu)
- [Add Feeding](https://www.icloud.com/shortcuts/96e5c97113354a9fb11e6325f443ff26) by @azosnu
- [Start Timer](https://www.icloud.com/shortcuts/ecd626a6a08b427cb333290a7e444569)
- [Stop Timer](https://www.icloud.com/shortcuts/2f61fac844354f729bfb40c67df69955)
- @a-camacho<sup>[ref](https://github.com/orgs/babybuddy/discussions/300#discussioncomment-2751969)</sup>
  - [Toggle Sleep](https://www.icloud.com/shortcuts/f9c89483104a4ff98b1d8c2258a56790)
  - [Toggle Feeding](https://www.icloud.com/shortcuts/44055bf099da4b3bb66d5c5e9afa944c)
- @lenaxia<sup>[ref](https://github.com/babybuddy/babybuddy/discussions/300#discussioncomment-5171942)</sup>:
  - [Add Feeding](https://www.icloud.com/shortcuts/d01e007ec08843278a909b104809ff00)
  - [Sleeping Time](https://www.icloud.com/shortcuts/f41e5ddf305e4e239619d9b03314c77f) (forked from @a-camacho's Toggle Sleep)
  - [Pumping](https://www.icloud.com/shortcuts/cd9d542e32e145d29b38d739f2751cb9)
- [Add Feeding](https://www.icloud.com/shortcuts/0afd187e87d846b4999f230b4bf66c77) by @SpeakingOfBrad<sup>[ref](https://github.com/orgs/babybuddy/discussions/300#discussioncomment-6065837)</sup> (forked from @lenaxia's Add Feeding)

**Comment by cdubz** [comment]
Thanks for sharing, @kejadlen! I didn’t know you could do form POST operations like this with Shortcuts… it would be nice to have a full set of functionality available for this. I’ll probably play around with adding more at some point.

**Reply by piraterx** [reply]
Hey I know you said you will look into creating more. Did you get any chance, I am new dad and was looking for easy way to do as wife isn't super tech. I liked Diaper change @kejadlen made. perfect. Easy. Specially for feeding as that's matter most. Nap and other stuff aren't so much relevant!!

**Reply by saddle-gaudier-06** [reply]
@kejadlen I get the below error when I added the Stop Timer shortcut. My API token works to start new timers, so I’m not sure why I’m getting the below error message. 

{"detail":"Authentication credentials were not provided."}

**Comment by cdubz** [comment]
See also #388.

I haven't had time look in to bringing together more of these lately but I would like to.

**Comment by iosmit** [comment]
I also improved upon @kejadlen Diaper Change by adding a question about 💩 color if Solid 💩 is selected by the user. 

[Diaper Change](https://www.icloud.com/shortcuts/3e7baf6586ac43af987614f51a7bcb66)

**Reply by saddle-gaudier-06** [reply]
With your shortcut, I've added my domain and used the API key as my token. Though, when I run the shortcut, I get the following message.

{"time":["Date\/time can not be in the future."]}

Any idea what I could be doing wrong?

**Reply by iosmit** [reply]
Which shortcut are you using Add Diaper Change or Add Feeding? If it’s Add Feeding make sure there isn’t an entry where it’s less than 7 minutes from prior entry. I have put gap of 7 minutes per entry/feeding.

**Reply by saddle-gaudier-06** [reply]
I was using Add Diaper Change, and I selected Wet for the type. 

UPDATE: I solved the issue by subtracting 30 seconds from the Current Time.

**Comment by BorjaLL** [comment]
For stop baby buddy I had to add “/“ at the end of the post messages for the APIs to get it to work.

I also had to remove the s from HTTPS but that’s on me :)

**Reply by saddle-gaudier-06** [reply]
Wow! That solved the issue!! Thanks for the help!!

**Comment by saddle-gaudier-06** [comment]
Has anyone tried using the Add Diaper Change (with color support) or Add Feeding via the Apple Watch? I keep getting an error that says “Conversion Error - Get Dictionary Value failed because Shortcuts couldn’t convert Text to Dictionary.”

**Reply by thefirstnoe1** [reply]
Fixed this by removing the "Get value for Color in Color" step under the Choose from "Color List" step.

**Comment by a-camacho** [comment]
Hi guys, great work.

Could the "Add feeding" shortcut set the start time at the moment shortcut is started, and end time at time when shortcut is finalized ?

Thanks

**Comment by a-camacho** [comment]
Hi guys, I played a bit more with the Shortcuts App and got two new shortcuts if anyone wants to try them.
It is for `feedings` and `sleep` and I used some steps from the Shortcuts above.

They work differently as they are `toggle` buttons.
You use the shortcut for creating new `sleep` or `feeding` (0 seconds of length) and then you can use the same shortcut to stop them. It will edit last entry, updating `end` time to actual time.

Here they are : 

- Toggle Sleep : https://www.icloud.com/shortcuts/f9c89483104a4ff98b1d8c2258a56790 
- Toggle Feeding : https://www.icloud.com/shortcuts/44055bf099da4b3bb66d5c5e9afa944c

**Reply by renoirb** [reply]
Those are amazing!
They suppress the need for timers.
One toggle per activity.
Pumping would basically be the same too. Except we have to ask the quantity at tge end side

**Reply by renoirb** [reply]
Not a question. Clarifications .

We should document that URL is actually domain:port only. No protocol. No path(?).

I guess this assumes BabyBuddy is behind something like Caddy or another proxy with TLS certificate.

in my case I have "babybuddy.home.myowndomainname.org:1111".  That answers over TLS. And made the home network router forward 1111 to Caddy, that has an babybuddy.home.myowndomainname.org:1111 entry

**Comment by BorjaLL** [comment]
anyone has issues with apple watch?
I've tried [Add Diaper Change](https://www.icloud.com/shortcuts/3e7baf6586ac43af987614f51a7bcb66) (with color support for solids) by @azosnu
 and get a dictionary error on the watch but ok in the iphone/ipad

![image](https://user-images.githubusercontent.com/3930245/168553372-09894131-d47d-407a-8981-8bb79a97a934.jpeg)

It only happens when including solids

**Reply by saddle-gaudier-06** [reply]
I’ve had the same issue.

**Reply by feranwq** [reply]
try move two boolean dictionaries before the if statement

**Reply by kelbyc97** [reply]
I’m having the same issue. Could you elaborate on a fix?

**Reply by feranwq** [reply]
> I’m having the same issue. Could you elaborate on a fix?

try this https://www.icloud.com/shortcuts/2e3b98d238634905aa2a4d91eefb2e6f

**Reply by kelbyc97** [reply]
Thank you for the quick reply! Still receiving the same message, screenshot attached(error in red)!
![20230523_141326000_iOS](https://github.com/babybuddy/babybuddy/assets/63815839/e9fc1264-e281-4975-ba41-a8c6f69a917b)

**Reply by kelbyc97** [reply]
I should've specified as well - it works flawlessly if you choose wet, it's only when you choose both or solid and a color is involved that the error occurs! Thank you!

**Reply by thefirstnoe1** [reply]
Fixed this by removing the "Get value for Color in Color" step under the Choose from "Color List" step.

**Comment by jimbolimbo13** [comment]
@kejadlen , here’s a shortcut to add a pumping session: https://www.icloud.com/shortcuts/d1df7a99a90c413a983f81abf2cd5673

**Comment by lenaxia** [comment]
I've further built upon @a-camacho 's Toggle Feeding skill here: https://www.icloud.com/shortcuts/6a38e5c7f5e04690afb50fe7cbd8957f (updated on 2/14 14:29 to fix a time bug)

The updated shortcut removes the "Start/Stop" menu in favor of checking if the previous feeding was complete. If not, it wraps up the previous feeding and adds an updated end time to it. If the previous feeding is complete, then it starts a new one, asking for feeding type and method. This eliminates 1 button click every time its used.

Additionally, it is now smart about asking about what kind of food is being used. For instance, if delivery method is through left/right/both breasts, then we know that the food is breast milk, so it skips that menu, saving 1 click. If delivery method is a bottle, we know the only options are breast milk/fortified/formula. Solid foods won't be delivered though a bottle. Likewise, if parent/self fed then we know we are only doing solids or solid foods. This saves at least 1 more click.

Finally, when wrapping up, it will ask how much food was consumed, either mL or grams. However, if we're delivering through breasts, theres no way to know how much food was consumed, so it skips that and fills in zero.

In total, it can reduce the number of clicks per feeding by almost half when the conditions are right.

**Reply by cdubz** [reply]
Nice additions -- thanks for sharing!

**Reply by lenaxia** [reply]
I've returned with more updates. 

This new feeding time asks for the starting and ending weight of the bottle and does the math for you meaning you don't have to keep track of weights anymore. It also moves the asking of which breast to the finish because you often don't know if you'll use both or just one at the beginning. It includes error checking for negative weight (if siri mishears you) and adds notes indicating starting and ending weight for better debugging. 

https://www.icloud.com/shortcuts/d01e007ec08843278a909b104809ff00


Sleeping time also now auto check if a nap has started or not and acts accordingly requiring no input from the user. 

https://www.icloud.com/shortcuts/f41e5ddf305e4e239619d9b03314c77f

Pumping also now is smarter and asks you for the weight of the pumping jar and auto subtract that from the final weight to get how much milk was pumped. 

https://www.icloud.com/shortcuts/cd9d542e32e145d29b38d739f2751cb9

Haven't made any changes to diaper because it's served it's purpose well.

**Reply by lenaxia** [reply]
@kejadlen
can you update the links accordingly?

**Reply by kelbyc97** [reply]
Any updates on diaper changes with color support? Seems to be something I’m missing, but I receive an error that “Get Dictionary Value” failed because it could not convert text to dictionary.

**Reply by thefirstnoe1** [reply]
I too get the dictionary value error on my phone... however, my wife doesn't get the error on her phone. Same iOS version. Very odd.

**Reply by thefirstnoe1** [reply]
Fixed this by removing the "Get value for Color in Color" step under the Choose from "Color List" step.

**Comment by SpeakingOfBrad** [comment]
I’ve made some changes to @lenaxia's shortcut that just made sense to me.
Most of the changes just change the wording/flow to be more inline with that you see in the BabyBuddy Add a Feeding interface.
* Changed the order of questions asked. Feeding type is asked first, then feeding method. Still uses the logical assumptions so that it only asks you methods that could work with that food type.
* Removed seconds from the date/time.
    * The Add a Feeding interface doesn’t allow you to enter seconds, which meant the feedings that were manually entered looked different than the shortcut entered feedings.
* When entering how much food there is or was consumed, changed the wording from asking for “weight (in grams)” to just “amount”.
    * I don’t weigh my bottles, and I’m sure a lot of other people don’t either. I just measure the milk by volume/ounces.
    * Plus using just “amount” means I can use ounces for bottles, and grams for solid food.

I did also fix a couple of functional issues I had:
* Fixed the “Conversion Error” “Get Dictionary Value failed because Shortcuts couldn't convert from Text to Dictionary.” error.
    * There was a Get Dictionary Value that was totally redundant that was failing. So I just deleted it and the shortcut runs just fine without it.
* Instead of hardcoding the URL to use https, you now specify https or http when you enter your URL that you use to access your BabyBuddy.

The shortcut is here: [https://www.icloud.com/shortcuts/0afd187e87d846b4999f230b4bf66c77](https://www.icloud.com/shortcuts/0afd187e87d846b4999f230b4bf66c77)

A quick note that I’d like to mention because it would be easy to miss: feedings can be started and stopped from two (or more) separate devices. Which means you can start a feeding from your phone, and your partner can stop it from their phone! This is because of the (brilliant) way @a-camacho uses to start/stop the feeding by setting the duration to 0 at start, and then checking for a 0 duration the next time the shortcut is run.

**Comment by dav3shanahan** [comment]
Hi, is there a way to add your app functionality to the rabbit r1?

**Comment by acmill14** [comment]
Hi! This is my first time sharing a shortcut, so if it turns out horrible I apologize in advance!

I made some additions to the "Add a Feeding" shortcut from @SpeakingOfBrad as a prep for my baby to be born in November. Some changes are practical, others were just personal preference. Here are my changes:

**Breast / Bottle Feeding Changes**
- If it's been more than a set amount of time (default is 90 minutes, but you can adjust this in set up or with a variable at the top) since the last feeding started & it was never "ended" via the shortcut, the shortcut assumes the caregiver forgot to end it. The shortcut automatically ends the feeding by using a default feeding duration (this is also adjusted in setup) and leaves the starting amount as the final amount. This all happens automatically, without the user being prompted. After updating the last feeding, the shortcut restarts to track the next feeding (however this is the first prompt the user will see). Just be sure to double-check that the 'Run Shortcut' action is choosing the correct shortcut (again, this is my first one being shared so I'm not sure if you need to manually check this). The thought behind this was if I'm exhausted and forget to end a feeding, it seems easier than having to trigger the shortcut twice & have a random breastfeeding session logged as taking 3 hours.
- If the amount in a bottle at the end is more than the starting amount, instead of sending an error notification and ending the shortcut, it will assume it's a typo and default to 0 as the ending amount

**Solid Food Changes**, these are mostly personal preference
- This version of the shortcut assumes that you are inputting solid foods at the end of a feeding only. The end time is the current time, and start time is a set time before (the default feeding duration mentioned above). There is also a prompt that will ask what the baby ate, and this will be added to the feeding as a note

**Miscellaneous**
- I added the baby's name as a setup variable. This is just because I like having more personal prompts (such as "What did Jane eat?" instead of "What type of food")
- I adjusted some wording and added a few emojis, but not many (I'm trying to make my brain think as little as possible for these upcoming night time feedings!)

I'm thinking of adding a tag for "nighttime feeding" or something similar that will be auto-added, but we'll see. You may hear from me again soon.

Here ya go! [https://www.icloud.com/shortcuts/fbeb22ad5a284aa9a4e8aaee33314df3](https://www.icloud.com/shortcuts/fbeb22ad5a284aa9a4e8aaee33314df3)

**Reply by FondledYeti** [reply]
I've been struggling to figure out how to get an automation to set a tag based on input.  Did you ever follow through with your nighttime feeding tag?

**Reply by acmill14** [reply]
Hi! I did make tags in my diaper shortcut that I think work (haven't checked it recently) via an API post. Here's how that portion of the shortcut is set up! It took me a little to figure out how it wanted to format it![image](https://github.com/user-attachments/assets/b1bf6e8e-1186-4329-ba8a-38f831c794ff)

**Comment by klaptafel** [comment]
Been using Baby Buddy for my previous child and just updated for my Baby Buddy Shortcut(s) to work with the next. Although it exists of multiple, you use them all by running the main shortcut. It's using Home Assistant notifications (so your partner is up to date as well), but you could easily replace it if you don't use Home Assistant. 

Your shortcuts really helped shaping mine, thanks!

https://github.com/klaptafel/apple-shortcuts-babybuddy

**Reply by uklok-agent** [reply]
Hey @kejadlen, your shortcuts sound really interesting. However, I can’t test them because you haven’t shared them publicly.

> Error when trying to use them
<img width="261" height="225" alt="image" src="https://github.com/user-attachments/assets/f7ff871b-4897-4f1c-8dad-427baa0dea82" />
